torch/distributed/_tensor/_dispatch.py

```diff
@@ -356,6 +364,8 @@ def try_get_replicate_spec(
                 args_schema.append(arg)
                 local_args.append(arg)
 
+        tree_map(arg_to_spec, args)
```

Comment:
hmm trying to understand more on this, we do have pytree to flatten the args input for certain ops, i.e. the foreach op list have pytree flatten enabled https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/ops/_pointwise_ops.py#L641

Reply:
oh right, is the idea that DTensor wants to manually specify which ops actually need the pytree machinery, so their usage is more limited (to avoid the perf impact?)

In that case, we can probably tweak this code to only do the tree_map here for ops that have opted into "requiring pytrees" (probably by checking the flag that you just linked)

Reply:
> oh right, is the idea that DTensor wants to manually specify which ops actually need the pytree machinery, so their usage is more limited (to avoid the perf impact?)

Yep!

> In that case, we can probably tweak this code to only do the tree_map here for ops that have opted into "requiring pytrees" (probably by checking the flag that you just linked)

We already sorta doing that in a few lines above https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/_dispatch.py#L291-L294

So I am wondering if there're some additional bugs that triggered that issue

Reply:
oh you are totally right... let me take another look

Reply:
ah yep you're right. So:

(1) this logic does the right thing of operating on the flattened args, as long as we detect that the op's sharding rule has opted into using pytrees

(2) the only bug is that DTensor was missing sharding rules for a few foreach ops, causing DTensor to use the "default" path of not using pytrees, in a "bad" way that causes DTensor to infinite loop horrible (I think, before getting a chance to error about not having a proper sharding rule).

hmm @wanchaol - is there any easy way to know earlier in advance if an op has no sharding prop rule and thus will fail, so we can check it here? https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/_dispatch.py#L291

One option is maybe to detect when there's no sharding prop rule and error earlier, before we try to generate FakeTensor arguments.

Reply:
(for now, I just "fixed" the issue by updated the foreach_ops list that DTensor relies on)

Reply:
> ah yep you're right. So:
> 
> (1) this logic does the right thing of operating on the flattened args, as long as we detect that the op's sharding rule has opted into using pytrees
> 
> (2) the only bug is that DTensor was missing sharding rules for a few foreach ops, causing DTensor to use the "default" path of not using pytrees, in a "bad" way that causes DTensor to infinite loop horrible (I think, before getting a chance to error about not having a proper sharding rule).
> 
> hmm @wanchaol - is there any easy way to know earlier in advance if an op has no sharding prop rule and thus will fail, so we can check it here? https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/_dispatch.py#L291
> 
> One option is maybe to detect when there's no sharding prop rule and error earlier, before we try to generate FakeTensor arguments.

For (2), I think the issue is that `_propagate_tensor_meta(op_schema)` get called first and errors out so we could see all different errors (here are some more examples where we don't see the infinite loop but some other error: https://github.com/pytorch/pytorch/issues/124990) even though the true reason is that no shardping prop has been registered for an op.   
https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/_sharding_prop.py#L198